File: app/services/collab/delta.py
"""Workbench delta —— 组织态增量补丁的纯函数管线（Workbench V6 / ADR-0119）。

delta schema（**绝对值语义** —— 全部字段是目标态而非增减量，重放幂等）::

    {
      "setGroups":       [{"id": str, "name"?: str, "collapsed"?: bool,
                           "parentId"?: str|null}, ...]   # 部分字段更新；
                                                          # id 不存在 = 创建（要求 name）
      "removeGroupIds":  [str, ...]                      # 级联子孙 + membership 清空
      "membershipSet":   [{"layerId": str, "groupId": str}, ...]
      "membershipClear": [layerId, ...]
      "locksAdd":        [layerId, ...]
      "locksRemove":     [layerId, ...]
    }

应用管线（固定次序，评审 R1-M4 —— setGroups → removeGroupIds →
membershipSet → membershipClear → locks）：
1. setGroups：create / 部分字段 patch（parentId 必须引用当前已存在组）；
2. removeGroupIds：与 setGroups 同 id → 冲突；级联移除子孙组；
3. membershipSet：目标组必须此刻存在；同键 Clear 后 Set（Set 优先）；
4. locksAdd / locksRemove；
5. 结果 doc 重跑引擎全量校验（调用方负责）+ 盖 `_rev`（调用方负责）。

mode 不在 delta 域（V5 R1-M2 决策一致：mode 不参与组织态撤销/增量）。
本模块纯函数、无 IO；引擎与前端共享同一语义（前端 TypeScript 侧为
镜像实现，differential 测试锚定）。
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional, Tuple

#: delta 各列表域上限（patch 级；超出 = 400，不静默截断）。
MAX_SET_GROUPS = 2000
MAX_REMOVE_GROUPS = 2000
MAX_MEMBERSHIP_SET = 2000
MAX_MEMBERSHIP_CLEAR = 2000
MAX_LOCKS = 2000
#: delta 载荷字节上限（总线与持久化通道各自预算中的 delta 份额）。
MAX_DELTA_BYTES = 64 * 1024

_RESERVED_DOC_KEYS = frozenset({"version", "groups", "membership", "lockedLayerIds", "mode", "_rev"})


class DeltaError(ValueError):
    """delta 非法（结构/引用/冲突/超限）——调用方映射 400。"""


def validate_delta(delta: Any) -> Dict[str, Any]:
    """结构校验（不查引用）；非法抛 DeltaError。返回规范化后的 delta。"""
    if not isinstance(delta, dict):
        raise DeltaError("workbench delta must be an object.")
    unknown = set(delta) - {
        "setGroups", "removeGroupIds", "membershipSet", "membershipClear",
        "locksAdd", "locksRemove",
    }
    if unknown:
        raise DeltaError(f"workbench delta has unknown fields: {sorted(unknown)!r}.")
    if not any(delta.get(k) for k in delta):
        raise DeltaError("workbench delta is empty.")

    def _str_list(field: str) -> List[str]:
        value = delta.get(field)
        if value is None:
            return []
        if not isinstance(value, list) or not all(isinstance(x, str) and x for x in value):
            raise DeltaError(f"workbench delta.{field} must be a list of non-empty strings.")
        return value

    set_groups = delta.get("setGroups") or []
    if not isinstance(set_groups, list) or len(set_groups) > MAX_SET_GROUPS:
        raise DeltaError(f"workbench delta.setGroups must be a list ≤ {MAX_SET_GROUPS}.")
    norm_groups: List[Dict[str, Any]] = []
    for entry in set_groups:
        if not isinstance(entry, dict) or not isinstance(entry.get("id"), str) or not entry["id"]:
            raise DeltaError("workbench delta.setGroups entries require string id.")
        patch: Dict[str, Any] = {"id": entry["id"]}
        if "name" in entry:
            if not isinstance(entry["name"], str) or not entry["name"]:
                raise DeltaError("workbench delta.setGroups name must be a non-empty string.")
            patch["name"] = entry["name"][:200]
        if "collapsed" in entry:
            if not isinstance(entry["collapsed"], bool):
                raise DeltaError("workbench delta.setGroups collapsed must be boolean.")
            patch["collapsed"] = entry["collapsed"]
        if "parentId" in entry:
            if entry["parentId"] is not None and not isinstance(entry["parentId"], str):
                raise DeltaError("workbench delta.setGroups parentId must be a string or null.")
            patch["parentId"] = entry["parentId"]
        norm_groups.append(patch)

    remove_ids = _str_list("removeGroupIds")
    if len(remove_ids) > MAX_REMOVE_GROUPS:
        raise DeltaError(f"workbench delta.removeGroupIds ≤ {MAX_REMOVE_GROUPS}.")

    membership_set = delta.get("membershipSet") or []
    if not isinstance(membership_set, list) or len(membership_set) > MAX_MEMBERSHIP_SET:
        raise DeltaError(f"workbench delta.membershipSet ≤ {MAX_MEMBERSHIP_SET} entries.")
    norm_ms: List[Tuple[str, str]] = []
    for entry in membership_set:
        if (
            not isinstance(entry, dict)
            or not isinstance(entry.get("layerId"), str)
            or not entry["layerId"]
            or not isinstance(entry.get("groupId"), str)
            or not entry["groupId"]
        ):
            raise DeltaError("workbench delta.membershipSet entries require layerId+groupId strings.")
        norm_ms.append((entry["layerId"], entry["groupId"]))

    membership_clear = _str_list("membershipClear")
    if len(membership_clear) > MAX_MEMBERSHIP_CLEAR:
        raise DeltaError(f"workbench delta.membershipClear ≤ {MAX_MEMBERSHIP_CLEAR}.")

    locks_add = _str_list("locksAdd")
    locks_remove = _str_list("locksRemove")
    if len(locks_add) > MAX_LOCKS or len(locks_remove) > MAX_LOCKS:
        raise DeltaError(f"workbench delta locks ≤ {MAX_LOCKS}.")

    norm = {
        "setGroups": norm_groups,
        "removeGroupIds": remove_ids,
        "membershipSet": norm_ms,
        "membershipClear": membership_clear,
        "locksAdd": locks_add,
        "locksRemove": locks_remove,
    }
    # R1-M4 预算兑现：patch 自身 ≤64KB（防请求体放大；全量 doc 由 256KB 闸管）。
    import json as _json

    try:
        encoded = len(_json.dumps(norm, ensure_ascii=False).encode("utf-8"))
    except (TypeError, ValueError):
        raise DeltaError("workbench delta is not JSON-serializable.") from None
    if encoded > MAX_DELTA_BYTES:
        raise DeltaError(f"workbench delta exceeds {MAX_DELTA_BYTES // 1024}KB; split the patch.")
    return norm


def _descendants(groups_by_id: Dict[str, Dict[str, Any]], root_ids: set[str]) -> set[str]:
    """子孙组 id（含自身）；parent 缺失按根处理（与投影/归一化语义一致）。"""
    children: Dict[str, List[str]] = {}
    for gid, g in groups_by_id.items():
        parent = g.get("parentId")
        if isinstance(parent, str) and parent in groups_by_id:
            children.setdefault(parent, []).append(gid)
    out: set[str] = set()
    stack = list(root_ids)
    while stack:
        cur = stack.pop()
        if cur in out:
            continue
        out.add(cur)
        stack.extend(children.get(cur, []))
    return out


def apply_delta(doc: Optional[Dict[str, Any]], delta: Dict[str, Any]) -> Dict[str, Any]:
    """把（已 validate 的）delta 应用到当前 doc → 新 doc（COW，不改入参）。

    结构性非法（引用不存在组、set+remove 同 id 等）抛 DeltaError —— 调用方
    映射 400（无半更新）。应用结果仍需引擎 `_workbench_doc_error` 全量校验
    （深度/环/体积在结果级复核 —— 部分字段 patch 可能造出超深链）。
    """
    base: Dict[str, Any] = dict(doc) if isinstance(doc, dict) else {}
    groups: List[Dict[str, Any]] = [dict(g) for g in base.get("groups", []) or [] if isinstance(g, dict)]
    membership: Dict[str, str] = {
        str(k): str(v) for k, v in (base.get("membership", {}) or {}).items()
    }
    locked: List[str] = [str(x) for x in (base.get("lockedLayerIds", []) or [])]

    groups_by_id: Dict[str, Dict[str, Any]] = {g["id"]: g for g in groups}
    version = base.get("version", 5)
    mode = base.get("mode", "explore")

    # 1. setGroups：create / 部分字段 patch。
    for patch in delta["setGroups"]:
        gid = patch["id"]
        existing = groups_by_id.get(gid)
        if existing is None:
            if "name" not in patch:
                raise DeltaError(f"workbench delta: group {gid!r} does not exist; create requires name.")
            collapsed = patch.get("collapsed")
            node = {
                "id": gid,
                "name": patch.get("name", gid),
                "collapsed": bool(collapsed) if collapsed is not None else False,
                "parentId": patch.get("parentId"),
            }
            groups.append(node)
            groups_by_id[gid] = node
        else:
            for field in ("name", "collapsed", "parentId"):
                if field in patch:
                    existing[field] = patch[field]

    # 2. removeGroupIds：set+remove 同 id 冲突；级联子孙 + membership 清空。
    remove_ids = set(delta["removeGroupIds"])
    conflict = remove_ids & {p["id"] for p in delta["setGroups"]}
    if conflict:
        raise DeltaError(f"workbench delta: setGroups and removeGroupIds collide on {sorted(conflict)!r}.")
    if remove_ids:
        cascade = _descendants(groups_by_id, remove_ids)
        groups = [g for g in groups if g["id"] not in cascade]
        for gid in cascade:
            groups_by_id.pop(gid, None)
        if membership:
            membership = {lid: gid for lid, gid in membership.items() if gid not in cascade}

    # 3. membershipSet（目标组必须此刻存在）→ membershipClear（Set 优先）。
    for layer_id, gid in delta["membershipSet"]:
        if gid not in groups_by_id:
            raise DeltaError(f"workbench delta: membershipSet target group {gid!r} does not exist.")
        membership[layer_id] = gid
    cleared = set(delta["membershipClear"]) - {lid for lid, _ in delta["membershipSet"]}
    for layer_id in cleared:
        membership.pop(layer_id, None)

    # 4. locks。
    lock_set = set(locked)
    lock_set.update(delta["locksAdd"])
    lock_set.difference_update(delta["locksRemove"])
    locked = list(lock_set)

    return {
        "version": version,
        "groups": groups,
        "membership": membership,
        "lockedLayerIds": locked,
        "mode": mode if isinstance(mode, str) else "explore",
    }


def doc_workbench_revision(doc: Optional[Dict[str, Any]]) -> Optional[int]:
    """读取 doc 的 `_rev` 盖章（无盖章的旧 doc 返回 None）。"""
    if not isinstance(doc, dict):
        return None
    rev = doc.get("_rev")
    if isinstance(rev, int) and rev >= 0:
        return rev
    return None

File: frontend/lib/workbench/delta.ts
/**
 * Workbench delta —— 组织态增量补丁纯函数（Workbench V6 / ADR-0119）。
 *
 * 与后端 `app/services/collab/delta.py` 语义镜像（differential 测试锚定）：
 * 绝对值语义（重放幂等）、固定应用次序、部分字段组更新、引用存在性检查。
 * undo 反演 = inverse delta（field 级），不再整表快照回写（R1-M3/M5）。
 */
import type { WorkbenchDocV5 } from './doc';

/** 组部分字段补丁（缺省键不动；id 不存在 = 创建，要求 name）。 */
export interface WorkbenchGroupPatch {
  id: string;
  name?: string;
  collapsed?: boolean;
  parentId?: string | null;
}

export interface MembershipSetEntry {
  layerId: string;
  groupId: string;
}

/** delta 各列表域上限（与后端同值；超出 = 拒绝提交，回退全量路径）。 */
export const MAX_SET_GROUPS = 2000;
export const MAX_REMOVE_GROUPS = 2000;
export const MAX_MEMBERSHIP_SET = 2000;
export const MAX_MEMBERSHIP_CLEAR = 2000;
export const MAX_LOCKS = 2000;

export interface WorkbenchDelta {
  setGroups?: WorkbenchGroupPatch[];
  removeGroupIds?: string[];
  membershipSet?: MembershipSetEntry[];
  membershipClear?: string[];
  locksAdd?: string[];
  locksRemove?: string[];
}

interface GroupsById {
  [id: string]: { id: string; name: string; collapsed: boolean; parentId: string | null };
}

function groupsByIdOf(doc: WorkbenchDocV5): GroupsById {
  const map: GroupsById = {};
  for (const g of doc.groups) map[g.id] = { ...g };
  return map;
}

/** 子孙组 id（含自身）；parent 缺失按根处理（与投影/归一化语义一致）。 */
function descendantsOf(groupsById: GroupsById, roots: Set<string>): Set<string> {
  const children: Record<string, string[]> = {};
  for (const [gid, g] of Object.entries(groupsById)) {
    if (g.parentId != null && groupsById[g.parentId]) {
      (children[g.parentId] ??= []).push(gid);
    }
  }
  const out = new Set<string>();
  const stack = [...roots];
  while (stack.length > 0) {
    const cur = stack.pop()!;
    if (out.has(cur)) continue;
    out.add(cur);
    for (const child of children[cur] ?? []) stack.push(child);
  }
  return out;
}

export class WorkbenchDeltaError extends Error {}

/**
 * 应用 delta（已 validate 的形状）→ 新 doc（不改入参）。
 * 结构性非法（引用不存在组、set+remove 同 id）抛 WorkbenchDeltaError。
 */
export function applyWorkbenchDelta(doc: WorkbenchDocV5, delta: WorkbenchDelta): WorkbenchDocV5 {
  const groupsById = groupsByIdOf(doc);

  // 1. setGroups：create / 部分字段 patch。
  for (const patch of delta.setGroups ?? []) {
    const existing = groupsById[patch.id];
    if (existing == null) {
      if (patch.name == null || patch.name === '') {
        throw new WorkbenchDeltaError(`group ${patch.id} does not exist; create requires name.`);
      }
      groupsById[patch.id] = {
        id: patch.id,
        name: patch.name,
        collapsed: patch.collapsed ?? false,
        parentId: patch.parentId ?? null,
      };
    } else {
      if (patch.name !== undefined) existing.name = patch.name;
      if (patch.collapsed !== undefined) existing.collapsed = patch.collapsed;
      if (patch.parentId !== undefined) existing.parentId = patch.parentId;
    }
  }

  // 2. removeGroupIds：set+remove 同 id 冲突；级联子孙 + membership 清空。
  const removeIds = new Set(delta.removeGroupIds ?? []);
  for (const patch of delta.setGroups ?? []) {
    if (removeIds.has(patch.id)) {
      throw new WorkbenchDeltaError(`setGroups and removeGroupIds collide on ${patch.id}.`);
    }
  }
  let removedGroups: Set<string> = new Set();
  if (removeIds.size > 0) {
    removedGroups = descendantsOf(groupsById, removeIds);
    for (const gid of removedGroups) delete groupsById[gid];
  }

  // 3. membershipSet（目标组必须此刻存在）→ membershipClear（Set 优先）。
  // 只清指向**本次被移除组**的键（与后端管线一致 —— 既有悬空键透传容忍）。
  const membership: Record<string, string> = {};
  for (const [layerId, groupId] of Object.entries(doc.membership)) {
    if (removedGroups.has(groupId)) continue;
    membership[layerId] = groupId;
  }
  for (const { layerId, groupId } of delta.membershipSet ?? []) {
    if (!groupsById[groupId]) {
      throw new WorkbenchDeltaError(`membershipSet target group ${groupId} does not exist.`);
    }
    membership[layerId] = groupId;
  }
  const setIds = new Set((delta.membershipSet ?? []).map((m) => m.layerId));
  for (const layerId of delta.membershipClear ?? []) {
    if (!setIds.has(layerId)) delete membership[layerId];
  }

  // 4. locks（保序：先 add 后 remove 的净效果，集合语义）。
  const lockSet = new Set(doc.lockedLayerIds);
  for (const id of delta.locksAdd ?? []) lockSet.add(id);
  for (const id of delta.locksRemove ?? []) lockSet.delete(id);

  return {
    version: 5,
    groups: Object.values(groupsById),
    membership,
    lockedLayerIds: [...lockSet],
    mode: doc.mode,
  };
}

/**
 * 结构化 diff：before → after 的最小 delta（O(n)）。
 * 无差异返回 null（零提交）。diff 失败/超限由调用方回退全量路径。
 * 非法变更（如 membership 指向不存在的组）返回 null（宁可全量也不要非法 delta）。
 */
export function diffWorkbenchDocs(before: WorkbenchDocV5, after: WorkbenchDocV5): WorkbenchDelta | null {
  const delta: WorkbenchDelta = {};
  const groupIdsBefore = new Set(before.groups.map((g) => g.id));
  const groupIdsAfter = new Set(after.groups.map((g) => g.id));
  const beforeById = new Map(before.groups.map((g) => [g.id, g]));

  // 组：改字段（before 有）+ 新建（before 无）。
  const setGroups: WorkbenchGroupPatch[] = [];
  for (const g of after.groups) {
    const prev = beforeById.get(g.id);
    if (prev == null) {
      setGroups.push({ id: g.id, name: g.name, collapsed: g.collapsed, parentId: g.parentId });
    } else if (
      prev.name !== g.name || prev.collapsed !== g.collapsed || prev.parentId !== g.parentId
    ) {
      const patch: WorkbenchGroupPatch = { id: g.id };
      if (prev.name !== g.name) patch.name = g.name;
      if (prev.collapsed !== g.collapsed) patch.collapsed = g.collapsed;
      if (prev.parentId !== g.parentId) patch.parentId = g.parentId;
      setGroups.push(patch);
    }
  }
  if (setGroups.length > 0) delta.setGroups = setGroups;
  if (setGroups.length > MAX_SET_GROUPS) return null;

  // 删除的组。
  const removeGroupIds = [...groupIdsBefore].filter((id) => !groupIdsAfter.has(id));
  if (removeGroupIds.length > 0) delta.removeGroupIds = removeGroupIds;
  if (removeGroupIds.length > MAX_REMOVE_GROUPS) return null;

  // membership：键级增改删。
  const membershipSet: MembershipSetEntry[] = [];
  const membershipClear: string[] = [];
  for (const [layerId, groupId] of Object.entries(after.membership)) {
    if (before.membership[layerId] !== groupId) membershipSet.push({ layerId, groupId });
  }
  for (const layerId of Object.keys(before.membership)) {
    if (after.membership[layerId] == null) membershipClear.push(layerId);
  }
  if (membershipSet.length > 0) delta.membershipSet = membershipSet;
  if (membershipClear.length > 0) delta.membershipClear = membershipClear;
  if (membershipSet.length > MAX_MEMBERSHIP_SET || membershipClear.length > MAX_MEMBERSHIP_CLEAR) return null;

  // 锁：集合差。
  const beforeLocks = new Set(before.lockedLayerIds);
  const afterLocks = new Set(after.lockedLayerIds);
  const locksAdd = after.lockedLayerIds.filter((id) => !beforeLocks.has(id));
  const locksRemove = before.lockedLayerIds.filter((id) => !afterLocks.has(id));
  if (locksAdd.length > 0) delta.locksAdd = locksAdd;
  if (locksRemove.length > 0) delta.locksRemove = locksRemove;
  if (locksAdd.length > MAX_LOCKS || locksRemove.length > MAX_LOCKS) return null;

  // mode 不在 delta 域（V5 R1-M2）；delta 覆盖不了 mode 变更 → 由全量路径处理。
  if (before.mode !== after.mode) return null;

  if (Object.keys(delta).length === 0) return null;

  // 合法性预检：diff 出的 membershipSet 必须指向 after 中存在的组（否则
  // 服务端会 400 —— 此时回退全量让服务端校验兜底语义保持一致）。
  for (const { groupId } of membershipSet) {
    if (!groupIdsAfter.has(groupId)) return null;
  }
  // 新建组的 parentId 必须指向 after 中存在的组（顺序无关的目标态）。
  for (const patch of setGroups) {
    if (patch.parentId != null && !groupIdsAfter.has(patch.parentId)) return null;
  }
  return delta;
}

/**
 * inverse delta：撤销 forward 在 before → after 上的效果（field 级）。
 * 构造自捕获的 forward delta 与前置 doc —— 并发下只回退本命令触碰的字段
 * （R1-M5：不再整表快照回写）。不可构造时抛错（调用方 journal-only）。
 */
export function invertWorkbenchDelta(
  forward: WorkbenchDelta,
  beforeDoc: WorkbenchDocV5,
): WorkbenchDelta {
  const inverse: WorkbenchDelta = {};

  // 反演 locks。
  if (forward.locksAdd?.length) inverse.locksRemove = [...forward.locksAdd];
  if (forward.locksRemove?.length) inverse.locksAdd = [...forward.locksRemove];

  // 反演 membership：Set → 恢复 before 值（before 无该键 → Clear）；
  // Clear → 恢复 before 值。forward 未触碰的键不动。
  const membershipSetBack: MembershipSetEntry[] = [];
  const membershipClearBack: string[] = [];
  for (const { layerId } of forward.membershipSet ?? []) {
    const prev = beforeDoc.membership[layerId];
    if (prev == null) membershipClearBack.push(layerId);
    else membershipSetBack.push({ layerId, groupId: prev });
  }
  for (const layerId of forward.membershipClear ?? []) {
    const prev = beforeDoc.membership[layerId];
    if (prev != null) membershipSetBack.push({ layerId, groupId: prev });
  }
  if (membershipSetBack.length > 0) inverse.membershipSet = membershipSetBack;
  if (membershipClearBack.length > 0) inverse.membershipClear = membershipClearBack;

  // 反演组：被删的 → 恢复 before 全节点（含被级联删的子孙）；被建的 → 删除；
  // 被改的 → 只回被改的字段。
  const createdIds = new Set<string>();
  const removedTrees: WorkbenchGroupPatch[] = [];
  const patchedBack: WorkbenchGroupPatch[] = [];
  const beforeById = new Map(beforeDoc.groups.map((g) => [g.id, g]));

  for (const patch of forward.setGroups ?? []) {
    const prev = beforeById.get(patch.id);
    if (prev == null) {
      createdIds.add(patch.id); // forward 创建 → inverse 删除
    } else {
      const back: WorkbenchGroupPatch = { id: patch.id };
      if (patch.name !== undefined && patch.name !== prev.name) back.name = prev.name;
      if (patch.collapsed !== undefined && patch.collapsed !== prev.collapsed) back.collapsed = prev.collapsed;
      if (patch.parentId !== undefined && patch.parentId !== prev.parentId) back.parentId = prev.parentId;
      if (back.name !== undefined || back.collapsed !== undefined || back.parentId !== undefined) {
        patchedBack.push(back);
      }
    }
  }
  for (const id of forward.removeGroupIds ?? []) {
    // 级联删除需恢复整个被删子树（before 中的子孙闭包）。
    const stack = [id];
    const seen = new Set<string>();
    while (stack.length > 0) {
      const cur = stack.pop()!;
      if (seen.has(cur)) continue;
      seen.add(cur);
      const prev = beforeById.get(cur);
      if (prev == null) continue;
      removedTrees.push({ id: prev.id, name: prev.name, collapsed: prev.collapsed, parentId: prev.parentId });
      // before 中以 cur 为父的组也在被删闭包内（与后端级联语义一致）。
      for (const g of beforeDoc.groups) {
        if (g.parentId === cur && !seen.has(g.id)) stack.push(g.id);
      }
    }
  }

  // 删除组的 inverse 先于字段回补（保证 parentId 引用存在性：先建组再挂父）。
  if (removedTrees.length > 0) inverse.setGroups = removedTrees;
  if (patchedBack.length > 0) (inverse.setGroups ??= []).push(...patchedBack);
  if (createdIds.size > 0) inverse.removeGroupIds = [...createdIds];

  // inverse 中 membershipSet 的目标组必须在 forward 前的 doc 中存在（不然
  // inverse 无法构造 —— 调用方捕获后 journal-only）。
  for (const m of inverse.membershipSet ?? []) {
    if (!beforeDoc.groups.some((g) => g.id === m.groupId)) {
      throw new WorkbenchDeltaError(`cannot invert: group ${m.groupId} missing in prior doc.`);
    }
  }

  if (Object.keys(inverse).length === 0) {
    throw new WorkbenchDeltaError('nothing to invert.');
  }
  return inverse;
}
